pkg/controller/pcideviceclaim/pcideviceclaim_controller.go

```diff
@@ -100,118 +183,290 @@ func unbindPCIDeviceFromVfioPCIDriver(addr string) error {
+		if isVfioPci && isOnThisNode && !pciDeviceIsClaimed(&pd, pdcs) {
 			logrus.Infof("PCI Device %s is bound to vfio-pci but has no Claim, attempting to unbind", pd.Status.Address)
-			err = unbindPCIDeviceFromVfioPCIDriver(pd.Status.Address)
+			err := unbindDeviceFromDriver(pd.Status.Address, vfioPCIDriver)
 			if err != nil {
+				logrus.Errorf("Error unbinding device from vfio-pci: %s", err)
 				return err
 			}
 		}
-		// After reboot, the PCIDeviceClaim will be there but the PCIDevice won't be bound to vfio-pci
-		if pd.Status.KernelDriverInUse != "vfio-pci" && nodename == pd.Status.NodeName {
-			// Set PassthroughEnabled to false
-			for _, pdc := range pdcs.Items {
-				if pdc.Spec.Address == pd.Status.Address {
-					logrus.Infof("Passthrough disabled for device %s", pd.Name)
-					pdc.Status.PassthroughEnabled = false
-					err = unbindPCIDeviceFromDriver(pd.Status.Address, pd.Status.KernelDriverInUse)
-					if err != nil {
-						logrus.Errorf("Error unbinding device after reboot: %s", err)
-					}
-					err = addNewIdToVfioPCIDriver(pd.Status.VendorId, pd.Status.DeviceId)
-					if err != nil {
-						logrus.Errorf("Error rebinding device after reboot: %s", err)
-					} else {
-						pdc.Status.PassthroughEnabled = true
-					}
-				}
-			}
+	}
+	return nil
+}
+
+// After reboot, the PCIDeviceClaim will be there but the PCIDevice won't be bound to vfio-pci
+func (h Handler) rebindAfterReboot(nodename string) {
+	logrus.Infof("Rebinding after reboot on node: %s", nodename)
+	pdcs, err := h.pdcClient.List(metav1.ListOptions{})
+	if err != nil {
+		logrus.Errorf("Error getting claims: %s", err)
+		return
+	}
+	for _, pdc := range pdcs.Items {
+		if pdc.Spec.NodeName != nodename {
+			continue
+		}
+		// Get PCIDevice for the PCIDeviceClaim
+		name := pdc.OwnerReferences[0].Name
+		pd, err := h.pdClient.Get(name, metav1.GetOptions{})
+		if err != nil {
+			logrus.Errorf("Error getting claim's device: %s", err)
+			continue
+		}
+
+		if pd.Status.KernelDriverInUse == "vfio-pci" {
+			logrus.Infof("PCIDevice %s is already bound to vfio-pci, skipping", pd.Name)
+			continue
+		}
+
+		logrus.Infof("Passthrough disabled for device %s", pd.Name)
+		pdcCopy := pdc.DeepCopy()
+
+		// Try to unbind from existing driver, if applicable
+		err = unbindDeviceFromDriver(pd.Status.Address, pd.Status.KernelDriverInUse)
+		if err != nil {
+			pdcCopy.Status.PassthroughEnabled = true
+			logrus.Errorf("Error unbinding device after reboot: %s", err)
+		} else {
+			pdcCopy.Status.PassthroughEnabled = false
+		}
+
+		// Enable Passthrough on the device
+		err = h.enablePassthrough(pd)
+		if err != nil {
+			logrus.Errorf("Error rebinding device after reboot: %s", err)
+			pdcCopy.Status.PassthroughEnabled = false
+		} else {
+			pdcCopy.Status.PassthroughEnabled = true
+		}
+		_, err = h.pdcClient.UpdateStatus(pdcCopy)
+		if err != nil {
+			logrus.Errorf("Failed to update PCIDeviceClaim status for %s: %s", pdc.Name, err)
```

Comment:
Should we return the error?

Reply:
@ibrokethecloud will help address this in a new PR.

Reply:
Checks out: https://github.com/harvester/pcidevices/pull/20/files